### PR TITLE
Refactor ExecuteTrtEngine.

### PR DESCRIPTION
**Pull request by tfeher**
This PR refactors [ExecuteTrtEngine](https://github.com/tensorflow/tensorflow/blob/34b86e53d7faaa62a9b62946a3a0a6d65c517eba/tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc#L653-L867) so that its components can be reused from [BuildAndRun](https://github.com/tensorflow/tensorflow/blob/34b86e53d7faaa62a9b62946a3a0a6d65c517eba/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L1287-L1353).

BuildAndRun is used by the TF-TRT op converter tests to initialize TRT bindings and execute inference. Its role is roughly equivalent to ExecuteTrtEngine. Recently ExecuteTrtEngine was modified to handle inference using dynamic shapes (PRs #36379, #36434, #36435), the same was not done for the opconverter's BuildAndRun. To allow unit testing of op converters in dynamic shape mode, one needs the extend BuildAndRun with similar changes.

This PR refactors ExecuteTRTEngine, so that both BuildAndRun and ExecuteTrtEngie can use a common set of utility functions to run inference. While doing these changes, the logging is streamlined: we return a status with the log message if an error occurs.  

For easier comparison, one could do something like:
```
git show master:tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc | sed -n '619,868p' > ExecuteTrtEngine_master.cc && diff ExecuteTrtEngine_master.cc tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc
```

This PR does not add new tests, nor does it modify any existing behavior. This PR is a prerequisite of op converter tests in dynamic shape mode. 

An alternative solution to this PR would be to copy the relevant code to BuildAndRun, but this would further increase the already existing code duplication between ExecuteTrtEngine and BuildAndRun. Furthermore, the exact way of specifying the input shapes is subject to change (for example currently we do not specify [input shape bindings](https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_execution_context.html#a306a7c94521d6a211be715ceb07c8c80), and it might be required in the future). Therefore it is preferred to reuse code between BuildAndRun and ExecuteTrtEngine.

Tagging @DEKHTIARJonathan and @bixia1 for review.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+// Gets the binding index of a tensor in an engine.
+//
+// The binding index is looked up using the tensor's name and the profile index.
+// Profile index should be set to zero, if we do not have optimization profiles.
+Status GetTrtBindingIndex(const char* tensor_name, int profile_index,
+                          const nvinfer1::ICudaEngine* cuda_engine,
+                          int* binding_index);
+
+// Set input buffers for TRT from a list of input tensors. The input tensors
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.h:56]: Sets

**Review by bixia1 — COMMENTED** [review]
Looks good to me. Thanks!

File: tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.h (hunk 1 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.h:56]: corrected

**Review by bixia1 — DISMISSED** [review]

**Comment by gbaned** [comment]
@tfeher Can you please address Ubuntu Sanity errors? Thanks!

**Comment by tfeher** [comment]
Fixed the sanity error.

**Review by bixia1 — DISMISSED** [review]

**Comment by bixia1** [comment]
The merge process is blocked due to the error below, can you please fix this?
tensorflow/compiler/tf2tensorrt/convert/utils.cc:169:3: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
  TensorShapeUtils::MakeShape(trt_dims.data(), trt_dims.size(), &shape);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tensorflow/compiler/tf2tensorrt/convert/utils.cc:179:3: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
  TensorShapeUtils::MakeShape(trt_dims.d, trt_dims.nbDims, &shape);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.

**Comment by tfeher** [comment]
Thanks @bixia1, I have fixed the problems with the unused results.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**